django/forms/fields.py

```diff
@@ -314,7 +314,9 @@ def __init__(self, *, max_value=None, min_value=None, step_size=None, **kwargs):
         if min_value is not None:
             self.validators.append(validators.MinValueValidator(min_value))
         if step_size is not None:
-            self.validators.append(validators.StepValueValidator(step_size))
+            self.validators.append(
+                validators.StepValueValidator(step_size, offset=min_value)
```

Comment:
This change is not tested, I'd add:
```diff
diff --git a/tests/forms_tests/field_tests/test_floatfield.py b/tests/forms_tests/field_tests/test_floatfield.py
index 276520f04d..61f1caf4cb 100644
--- a/tests/forms_tests/field_tests/test_floatfield.py
+++ b/tests/forms_tests/field_tests/test_floatfield.py
@@ -84,6 +84,15 @@ class FloatFieldTest(FormFieldAssertionsMixin, SimpleTestCase):
         self.assertEqual(-1.26, f.clean("-1.26"))
         self.assertEqual(f.step_size, 0.02)
 
+    def test_floatfield_step_size_min_value(self):
+        f = FloatField(step_size=0.02, min_value=0.01)
+        msg = "'Ensure this value is a multiple of step size 0.02.'"
+        with self.assertRaisesMessage(ValidationError, msg):
+            f.clean("0.02")
+        self.assertEqual(f.clean("2.33"), 2.33)
+        self.assertEqual(f.clean("0.11"), 0.11)
+        self.assertEqual(f.step_size, 0.02)
+
     def test_floatfield_widget_attrs(self):
         f = FloatField(widget=NumberInput(attrs={"step": 0.01, "max": 1.0, "min": 0.0}))
         self.assertWidgetRendersTo(
diff --git a/tests/forms_tests/field_tests/test_integerfield.py b/tests/forms_tests/field_tests/test_integerfield.py
index 1361b5cc9e..25ccc2d0d8 100644
--- a/tests/forms_tests/field_tests/test_integerfield.py
+++ b/tests/forms_tests/field_tests/test_integerfield.py
@@ -126,6 +126,19 @@ class IntegerFieldTest(FormFieldAssertionsMixin, SimpleTestCase):
         self.assertEqual(12, f.clean("12"))
         self.assertEqual(f.step_size, 3)
 
+    def test_integerfield_step_size_min_value(self):
+        f = IntegerField(step_size=3, min_value=-1)
+        self.assertWidgetRendersTo(
+            f,
+            '<input name="f" min="-1" step="3" type="number" id="id_f" required>',
+        )
+        msg = "'Ensure this value is a multiple of step size 3.'"
+        with self.assertRaisesMessage(ValidationError, msg):
+            f.clean("9")
+        self.assertEqual(f.clean("2"), 2)
+        self.assertEqual(f.clean("-1"), -1)
+        self.assertEqual(f.step_size, 3)
+
     def test_integerfield_localized(self):
         """

```

Reply:
Also, the current error message is quite confusing, e.g.
```
f = IntegerField(step_size=3, min_value=-1)
```
raises _"Ensure this value is a multiple of step size 3."_ when passing `6`.

Reply:
Yeah that's a good point felix.

Just throwing out ideas here but I'm wondering whether it'd almost be worth accounting for the most common use cases here - would that be even/odd numbers?  If so then you could tailor the error message appropriately here to "Please ensure this value is an even[odd] number".

Other than that what else could be done other than mentioning the offset/base with the most comprehensible wording?

> Ensure this value is a multiple of step size 3 with a base of 1

Reply:
Agreed that could be difficult to understand. When the validation is done client side with the value `2` for `<input type="number" min=1 step=2>` both Chrome and Firefox give the message `"Please enter a valid value. The two nearest valid values are 1 and 3."`. I think that is quite clear.

However, with that message I don't know how (or if) the min and/or max value should also be taken into account. If you enter a value of `-1` or `0` for that input then the browser only shows the message from the min value validation (`"Value must be greater than or equal to 1"`).

Reply:
@felixxm as proposed, I added 3 additional tests and changed the errors message for the `StepValueValidator` to the proposal of @jmsmkn .

Reply:
Although I like the nearest values, it adds quite a bit of complexity to generating the error message. I think it would be good to only slightly tweak from the original message. How do you feel about:

_"Please ensure this value is a multiple of step size 0.4, starting from 0.1, e.g. 0.1, 0.5, 0.9, and so on."_

@jrief I think this would also address @felixxm's concerns, what do you think?

Reply:
+1 to Sarah's proposal, the example helps considerably.

Reply:
Updated.